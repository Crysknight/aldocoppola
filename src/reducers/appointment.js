import { CHOOSE_CENTER, CHOOSE_EMPLOYEE, CONFIRM_SERVICES, CONFIRM_DATE_TIME } from '../actions/types';

export default (state = {}, action) => {
	switch (action.type) {
		case CHOOSE_CENTER: {
			let newState = JSON.stringify(state);
			newState = JSON.parse(newState);
			newState.centerChosen = action.payload;
			return newState;
		}
		case CHOOSE_EMPLOYEE: {
			let newState = JSON.stringify(state);
			newState = JSON.parse(newState);
			newState.employeeChosen = action.payload;
			return newState;
		}
		case CONFIRM_SERVICES: {
			let newState = JSON.stringify(state);
			newState = JSON.parse(newState);
			newState.servicesChosen = action.payload;
			return newState;
		}
		case CONFIRM_DATE_TIME: {
			let newState = JSON.stringify(state);
			newState = JSON.parse(newState);
			newState.dateTimeChosen = action.payload;
			return newState;
		}
		default: {
			return state;
		}
	}
};