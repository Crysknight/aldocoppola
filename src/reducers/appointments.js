import { 
	ADD_APPOINTMENT, 
	CANCEL_NEW_APPOINTMENT, 
	CHOOSE_CENTER,
	DELETE_APPOINTMENT,
	CONFIRM_APPOINTMENT_EDIT
} from '../actions/types';

export default (state = [], action) => {
	switch (action.type) {
		case ADD_APPOINTMENT: {
			let newState = JSON.stringify(state);
			newState = JSON.parse(newState);
			newState.push(action.payload);
			return newState;
		}
		case CANCEL_NEW_APPOINTMENT: {
			if (action.payload.isTheLastOneTheOnlyOne) {
				return [];
			} else {
				return state;
			}
		}
		case CHOOSE_CENTER: {
			return [];
		}
		case DELETE_APPOINTMENT: {
			let newState = JSON.stringify(state);
			newState = JSON.parse(newState);
			for (let i = 0; i < newState.length; i++) {
				if (newState[i].number === action.payload.appointment.number) {
					newState.splice(i, 1);
				}
			}
			for (let i = 0; i < newState.length; i++) {
				newState[i].number = i;
			}
			return newState;
		}
		case CONFIRM_APPOINTMENT_EDIT: {
			let newState = JSON.stringify(state);
			newState = JSON.parse(newState);
			for (let i = 0; i < newState.length; i++) {
				if (newState[i].number === action.payload.appointment.number) {
					delete action.payload.appointment.edited;
					newState[i] = action.payload.appointment;
				}
			}
			return newState;
		}
		default: {
			return state;
		}
	}
}