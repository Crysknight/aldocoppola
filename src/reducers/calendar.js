import moment from 'moment';
import 'moment/locale/ru';

import { CHOOSE_EMPLOYEE, CHOOSE_DATE, CHOOSE_TIME } from '../actions/types';

const getCalendar = (numberOfMonths) => {

	const calendar = [];

	for (let i = moment().month(); i < (moment().month() + numberOfMonths); i++) {

		let number = i % 12;
		let year = i < 12 ? moment().year() : moment().year() + 1;
		let name = moment([year, number]).format('MMMM');
		name = name.charAt(0).toUpperCase() + name.slice(1);
		let days = [];
		let numberOfDays = moment([year, number]).daysInMonth();
		for (let j = 1; j <= numberOfDays; j++) {
			let dayOfWeek = +moment([year, number, j]).format('d') - 1;
			let dayOfWeekName = moment([year, number, j]).format('dddd');
			dayOfWeekName = dayOfWeekName.charAt(0).toUpperCase() + dayOfWeekName.slice(1);
			let day = moment([year, number, j]);
			let today = moment([moment().year(), moment().month(), moment().date()]);
			days.push({
				date: j,
				dayOfWeek: dayOfWeek === -1 ? 6 : dayOfWeek,
				dayOfWeekName,
				past: moment(today).isAfter(day)
			});
		}
		calendar.push({
			number,
			name,
			year,
			days
		});

	}

	return calendar;

}

window.moment = moment;

export default (state = getCalendar(8), action) => {
	switch (action.type) {
		case CHOOSE_EMPLOYEE: {
			let newState = JSON.stringify(state);
			newState = JSON.parse(newState);
			for (let workHour of action.payload.workHoursFree) {
				workHour = moment(workHour, 'YYYY-MM-DD hh:mm');
				let month = workHour.month();
				let date = workHour.date();
				for (let calendarMonth of newState) {
					if (calendarMonth.number === month) {
						for (let calendarDate of calendarMonth.days) {
							if (calendarDate.date === date) {
								calendarDate.workHours = calendarDate.workHours ? calendarDate.workHours : {};
								let morningBorder = workHour.clone().hour(12).minute(0);
								let dayBorder = workHour.clone().hour(18).minute(0);
								if (workHour.isSameOrBefore(morningBorder)) {
									calendarDate.workHours["Утро"] = calendarDate.workHours["Утро"] ? calendarDate.workHours["Утро"] : [];
									calendarDate.workHours["Утро"].push(workHour.format('HH:mm'));
								} else if (workHour.isSameOrBefore(dayBorder)) {
									calendarDate.workHours["День"] = calendarDate.workHours["День"] ? calendarDate.workHours["День"] : [];
									calendarDate.workHours["День"].push(workHour.format('HH:mm'));
								} else {
									calendarDate.workHours["Вечер"] = calendarDate.workHours["Вечер"] ? calendarDate.workHours["Вечер"] : [];
									calendarDate.workHours["Вечер"].push(workHour.format('HH:mm'));
								}
								// calendarDate.workHours.push(workHour.format('HH:mm'));
							}
						}
					}
				}
			}
			return newState;
		}
		case CHOOSE_DATE: {
			let newState = JSON.stringify(state);
			newState = JSON.parse(newState);
			let month = action.payload.month;
			let date = action.payload.date;
			for (let calendarMonth of newState) {
				calendarMonth.chosen = calendarMonth.chosen ? false : calendarMonth.chosen;
				for (let calendarDate of calendarMonth.days) {
					calendarDate.chosen = calendarDate.chosen ? false : calendarDate.chosen;
					calendarDate.timeChosen = calendarDate.timeChosen ? null : calendarDate.timeChosen;
				}
				if (calendarMonth.number === month) {
					calendarMonth.chosen = true;
					for (let calendarDate of calendarMonth.days) {
						if (calendarDate.date === date) {
							calendarDate.chosen = true;
						}
					}
				}
			}
			return newState;
		}
		case CHOOSE_TIME: {
			let newState = JSON.stringify(state);
			newState = JSON.parse(newState);
			let month = action.payload.date.monthNumber;
			let date = action.payload.date.date;
			for (let calendarMonth of newState) {
				calendarMonth.chosen = calendarMonth.chosen ? false : calendarMonth.chosen;
				for (let calendarDate of calendarMonth.days) {
					calendarDate.chosen = calendarDate.chosen ? false : calendarDate.chosen;
					calendarDate.timeChosen = calendarDate.timeChosen ? null : calendarDate.timeChosen;
				}
				if (calendarMonth.number === month) {
					calendarMonth.chosen = true;
					for (let calendarDate of calendarMonth.days) {
						if (calendarDate.date === date) {
							calendarDate.chosen = true;
							calendarDate.timeChosen = action.payload.time;
						}
					}
				}
			}
			return newState;
		}
		default: {
			return state;
		}
	}
};