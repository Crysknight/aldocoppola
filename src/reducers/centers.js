import { GOT_CENTERS } from '../actions/types';

export default (state = [], action) => {
	switch (action.type) {
		case GOT_CENTERS: {
			return action.payload;
		}
		default: {
			return state;
		}
	}
};