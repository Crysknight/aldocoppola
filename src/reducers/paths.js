import { GOT_EMPLOYEES, GOT_MAIN_SERVICES } from '../actions/types';

// Basis of all widget paths (after this everything is widget)
const appPath = '/_appliance/';

// Regular expression to find any path related to the widget
const appRegExp = new RegExp(appPath.slice(0, -1) + '(.*)');

// Object with paths used by the app
const paths = {
	__app: {
		pathString: '',
		privacy: true
	},
	ChooseCenter: {
		pathString: 'choose-center/',
		privacy: false
	},
	ChooseEmployee: {
		pathString: 'choose-employee/',
		privacy: true
	},
	ChooseServices: {
		pathString: 'choose-services/',
		privacy: true,
		childPaths: {
			SubmitServices: {
				pathString: 'submit-services/',
				privacy: true
			}
		}
	},
	ChooseDateTime: {
		pathString: 'choose-datetime/',
		privacy: true
	}
};

const pathsMethods = {};

// Function for constructing pathstrings from the sctructure above
// It doesn't check the 'pathString' consistency so please regard the following format
// pathString: 'nameofthepathstring/' with slash at the end
pathsMethods.createPathStrings = (paths, nested) => {
	// Firstly, it loops through the paths object, finding properties that are paths too
	for (let path in paths) {
		// Do not process the root path of the widget
		if (path === '__app') {
			paths[path].pathString = appPath;
			continue;
		}
		let parentString;
		// The 'finished' property denotes that we already processed this pathstring
		// not to process it in further calls of function
		if (nested || paths[path].finished) {
			parentString = paths[path].pathString;
		} else {
			// Prepending 'appPath' at first level of structure
			parentString = paths[path].pathString = appPath + paths[path].pathString;
			paths[path].finished = true;
		}
		// If the path has childPaths, we process every childPath's pathString
		if (paths[path].childPaths) {
			let childPaths = paths[path].childPaths;
			for (let path in childPaths) {
				if (!childPaths[path].finished) {
					childPaths[path].pathString = parentString + childPaths[path].pathString;
					childPaths[path].finished = true;
				}
				// And recursively process the childPaths too
				pathsMethods.createPathStrings(childPaths, true);
			}
		}
	}
};

pathsMethods.createPathStrings(paths);

// A function to get the path of the specific page
pathsMethods.getPath = (paths, currentPath, path) => {
	if (!path) {
		return currentPath.replace(appRegExp, '') + paths.__app.pathString;
	}
	return currentPath.replace(appRegExp, '') + path.pathString;
};

// A function to create a Link that opens and closes the app
pathsMethods.getAppSwitch = (paths, currentPath) => {
	if (currentPath.match(appRegExp)) {
		return currentPath.replace(appRegExp, '');
	} else {
		return currentPath.replace(/\/$/, '') + paths.__app.pathString;
	}
};

// A function to create dynamic paths. 
// Takes in a path object: properties are - 'pathName', 'pathString', 'parentPath' and 'privacy'
pathsMethods.createPath = (paths, path) => {
	if (!path.pathString || !path.pathName) {
		throw new Error('pathString, or pathName properties missing');
	}
	if (path.pathString.match(/ /) !== null) {
		throw new Error('pathString should not contain spaces');
	}
	if (path.pathString.match(/^(\w|\d|-|_)*\/$/) === null) {
		throw new Error('pathString should begin from a letter, number, \'-\', or \'_\' and end with one slash symbol');
	}
	let targetPath = { 
		pathString: path.pathString,
		privacy: path.hasOwnProperty('privacy') ? path.privacy : true
	}
	// 'hasOwnProperty' because the 'parentPath' itself can be undefined
	if (!path.hasOwnProperty('parentPath')) {
		paths[path.pathName] = targetPath;
	} else if (path.hasOwnProperty('parentPath') && !path.parentPath) {
		throw new Error('Specified parentPath not found');
	} else {
		// createPath's recursive function to search a parentPath inside the paths object
		// In every path that has childPaths it also searches the parentPath in them recursively calling itself 
		let pathsCrawler = function(paths, targetPath, pathName, parentPath) {
			for (let path in paths) {
				if (paths[path].pathString === parentPath.pathString) {
					paths[path].childPaths = paths[path].childPaths ? paths[path].childPaths : {};
					paths[path].childPaths[pathName] = targetPath;
				} else if (paths[path].childPaths) {
					pathsCrawler(paths[path].childPaths, targetPath, pathName, parentPath);
				}
			}
		}
		pathsCrawler(paths, targetPath, path.pathName, path.parentPath);
	}
	pathsMethods.createPathStrings(paths);
};

// Export the paths and pathsMethods objects for testing purposes
export { pathsMethods };
export { paths };
// window.pathsMethods = pathsMethods;
// window.paths = paths;
// console.log(paths);

// Reducer, providing the paths object to the whole app
export default (state = paths, action) => {
	switch (action.type) {
		case GOT_EMPLOYEES: {
			let payload = action.payload;
			let newState = JSON.stringify(state);
			newState = JSON.parse(newState);
			for (let employee of payload) {
				let pathRecourse = 'employee' + employee.id;
				pathsMethods.createPath(newState, {
					pathName: pathRecourse,
					pathString: pathRecourse + '/',
					privacy: true,
					parentPath: newState.ChooseEmployee
				});
			}
			return newState;
		}
		case GOT_MAIN_SERVICES: {
			let payload = action.payload;
			let newState = JSON.stringify(state);
			newState = JSON.parse(newState);
			for (let service of payload) {
				let pathRecourse = 'service' + service.id;
				pathsMethods.createPath(newState, {
					pathName: pathRecourse,
					pathString: pathRecourse + '/',
					privacy: true,
					parentPath: newState.ChooseServices
				});
			}
			return newState;
		}
		default: {
			return state;
		}
	}
};