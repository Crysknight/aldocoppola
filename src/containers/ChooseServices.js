import React, { Component } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';

import actions from '../actions';

import Header from '../components/header';
import Service from '../components/service';

import SVGArrowLeft from '../components/svg-arrow-left';

import { pathsMethods } from '../reducers/paths';
import { mainServicesMethods} from '../reducers/mainServices';

class ChooseServices extends Component {

	// constructor(props) {
		// super(props);

	// }

	componentWillMount() {
		this.props.loadServices();
	}

	getMainService() {
		let paths = this.props.paths;
		let mainServices = this.props.mainServices;
		let regExp = new RegExp(paths.ChooseServices.pathString + 'service([\\w|\\d]*)\\/');
		let id = +this.props.match.path.replace(regExp, '$1');
		return mainServicesMethods.findMainServiceById(mainServices, id);
	}

	chooseService(id) {
		this.props.chooseService(id);
	}

	renderServices() {
		return this.props.services.map((service, index) => {
			return (
				<Service
					key={index}
					service={service}
					chooseService={() => this.chooseService(service.id)}
				/>
			);
		});
	}

	render() {
		let paths = this.props.paths;
		let mainService = this.getMainService();
		return (
			<div id="choose_services">
				<Header title={mainService.title}>
					<Link 
						to={pathsMethods.getPath(paths, this.props.match.path, paths.ChooseServices)}
						className="back-link"
					><SVGArrowLeft /></Link>
				</Header>
				{this.renderServices()}
			</div>
		);
	}

}

function mapStateToProps(state) {
	return {
		paths: state.paths,
		services: state.services,
		mainServices: state.mainServices
	};
}

function matchDispatchToProps(dispatch) {
	return bindActionCreators({
		loadServices: actions.loadServices,
		chooseService: actions.chooseService
	}, dispatch);
}

export default connect(mapStateToProps, matchDispatchToProps)(ChooseServices);