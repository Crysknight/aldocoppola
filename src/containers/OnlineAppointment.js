import React, { Component } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';

// import actions from '../actions';

import Header from '../components/header';
import Footer from '../components/footer';

import SVGLock from '../components/svg-lock';
import SVGArrowRight from '../components/svg-arrow-right';

import { pathsMethods } from '../reducers/paths';

class OnlineAppointment extends Component {

	// constructor(props) {
		// super(props);

	// }

	render() {
		let paths = this.props.paths;
		return (
			<div id="online_appointment">
				<Header title="Онлайн запись" />
				<Link 
					to={pathsMethods.getPath(paths, this.props.match.path, paths.ChooseCenter)}
					className="main-link chosen"
				>
					<span className="title">Изменить центр красоты</span>
					<span className="info-text">{this.props.appointment.centerChosen.name}</span>
					<SVGArrowRight />
				</Link>
				<Link 
					to={pathsMethods.getPath(paths, this.props.match.path, paths.ChooseEmployee)}
					className="main-link"
				>
					<span className="title">Выбрать сотрудника</span>
					<SVGArrowRight />
				</Link>
				<Link 
					to={pathsMethods.getPath(paths, this.props.match.path, paths.ChooseServices)}
					className="main-link"
				>
					<span className="title">Выбрать услуги</span>
					<SVGArrowRight />
				</Link>
				<Link 
					to={pathsMethods.getPath(paths, this.props.match.path, paths.ChooseDateTime)}
					className="main-link"
				>
					<span className="title">Выбрать дату и время</span>
					<SVGArrowRight />
				</Link>
				<Footer className="coal">
					<Link 
						to={pathsMethods.getPath(paths, this.props.match.path, paths.MyAccount)}
						className="footer-link my-account-link"
					><SVGLock />Личный кабинет</Link>
				</Footer>
			</div>
		);
	}

}

function mapStateToProps(state) {
	return {
		paths: state.paths,
		appointment: state.appointment
	};
}

function matchDispatchToProps(dispatch) {
	return bindActionCreators({
		
	}, dispatch);
}

export default connect(mapStateToProps, matchDispatchToProps)(OnlineAppointment);