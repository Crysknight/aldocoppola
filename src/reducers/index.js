import { combineReducers } from 'redux';
import paths from './paths';
import centers from './centers';
import appointment from './appointment';
import employees from './employees';

const allReducers = combineReducers({
	paths: paths,
	centers: centers,
	appointment: appointment,
	employees: employees
});

export default allReducers;
