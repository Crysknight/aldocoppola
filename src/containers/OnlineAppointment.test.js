import React from 'react';
import OnlineAppointment from './OnlineAppointment';
import renderer from 'react-test-renderer';
import { MemoryRouter, Route } from 'react-router-dom';
import { createStore, applyMiddleware } from 'redux';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import allReducers from '../reducers';
import { pathsMethods } from '../reducers/paths';
import { paths } from '../reducers/paths';
import ChooseServices from './ChooseServices';

describe('OnlineAppointment', () => {

	it('renders component properly', () => {
		const store = createStore(allReducers, applyMiddleware(thunk));
		store.dispatch({ type: 'CHOOSE_CENTER', payload: { name: 'Generic Name', id: 8 } });
		const component = renderer.create(
			<Provider store={store}>
				<MemoryRouter initialEntries={[ '/test/bauble.html' + paths.__app.pathString ]}>
					<Route path={'/test/bauble.html' + paths.__app.pathString} component={OnlineAppointment} />
				</MemoryRouter>
			</Provider>
		);
	});

});