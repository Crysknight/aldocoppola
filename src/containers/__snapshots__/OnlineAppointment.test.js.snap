// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OnlineAppointment renders component properly 1`] = `
<div
  id="online_appointment"
>
  <div
    className="ac-header"
  >
    <h3>
      Онлайн запись
    </h3>
  </div>
  <a
    className="main-link chosen"
    href="/test/bauble.html/_appliance/choose-center/"
    onClick={[Function]}
  >
    <span
      className="title"
    >
      Изменить центр красоты
    </span>
    <span
      className="info-text"
    >
      TEST_NAME
    </span>
  </a>
  <a
    className="main-link"
    href="/test/bauble.html/_appliance/choose-employee/"
    onClick={[Function]}
  >
    <span
      className="title"
    >
      Выбрать сотрудника
    </span>
  </a>
  <a
    className="main-link"
    href="/test/bauble.html/_appliance/choose-services/"
    onClick={[Function]}
  >
    <span
      className="title"
    >
      Выбрать услуги
    </span>
  </a>
  <a
    className="main-link"
    href="/test/bauble.html/_appliance/choose-datetime/"
    onClick={[Function]}
  >
    <span
      className="title"
    >
      Выбрать дату и время
    </span>
  </a>
  <div
    className="ac-footer coal"
  >
    <a
      className="footer-link my-account-link"
      href="/test/bauble.html/_appliance/"
      onClick={[Function]}
    >
      Личный кабинет
    </a>
  </div>
</div>
`;
